src/apps/virtio_net/virtio_net_dev.lua

```diff
@@ -0,0 +1,208 @@
+   uint32_t host_features;
+   uint32_t guest_features;
+   uint32_t queue_pfn;
+   uint16_t queue_num;
+   uint16_t queue_sel;
+   uint16_t queue_notify;
+   uint8_t status;
+   uint8_t isr;
+   uint16_t config_vector;
+   uint16_t queue_vector;
+]]
+
+VGdev = {}
+VGdev.__index = VGdev
+
+function VGdev:new(args)
+   local min_features = 0 -- C.VIRTIO_F_VERSION_1?
+   local want_features = C.VIRTIO_NET_F_CSUM + C.VIRTIO_NET_F_MAC
+   pci.unbind_device_from_linux (args.pciaddr)
+
+   local bar = open_bar(pci.path(args.pciaddr..'/resource0'), virtio_pci_bar0)
+
+   bar.status = 0 -- Reset device.
+   bar.status = bor(bar.status, 1) -- Acknowledge.
+   -- Check something.
+   bar.status = bor(bar.status, 2) -- Driver.
+   local features = bar.host_features
+   if debug then print('host_features', features) end
+   if band(features, min_features) ~= min_features then
+      bar.status = bor(bar.status, 128) -- Failure.
+      bar:close()
+      return nil, "doesn't provide minimum features"
+   end
+   if debug then print('set features to:', band(features, want_features)) end
+   bar.guest_features = band(features, want_features)
+   bar.status = bor(bar.status, 8) -- Features OK.
+   if debug then print('got features: ', bar.host_features, bar.guest_features) end
+   if band(bar.status, 8) ~= 8 then
+      bar.status = bor(bar.status, 128) -- Failure.
+      bar:close()
+      return nil, "feature set wasn't accepted by device"
+   end
+
+   if debug then print("enumerating queues...") end
+   local vqs = {}
+   for qn = 0, 16 do
+      bar.queue_sel = qn
+      local queue_size = bar.queue_num
+      if queue_size == 0 then break end
+
+      if debug then print(('queue %d: size: %d'):format(qn, queue_size)) end
+      local vring = gvring.allocate_vring(bar.queue_num)
+      vqs[qn] = vring
+      -- VIRTIO_PCI_QUEUE_ADDR_SHIFT
+      bar.queue_pfn = rshift(vring.vring_physaddr, 12)
+   end
+
+   if not(vqs[0] and vqs[1]) then
+      bar.status = bor(bar.status, 128) -- Failure.
+      bar:close()
+      return nil, "missing required virtqueues"
+   end
+
+   bar.status = bor(bar.status, 4) -- Driver OK.
+
+   return setmetatable({
+      bar = bar,
+      vqs = vqs,
+   }, self)
+end
+
+function VGdev:close()
+   for qn, _ in ipairs(self.vqs) do
+      self.bar.queue_sel = qn
+      self.bar.queue_pfn = 0
+   end
+   self.bar:close()
+end
+
+function VGdev:can_transmit()
+   return self.vqs[1]:can_add()
+end
+
+local pk_header = ffi.new([[
+   struct {
+      uint8_t flags;
+      uint8_t gso_type;
+      int16_t hdr_len;
+      int16_t gso_size;
+      int16_t csum_start;
+      int16_t csum_offset;
+      // int16_t num_buffers;    // only if MRG_RXBUF feature active
+   }
+]])
+
+function VGdev:transmit(p)
+   ffi.fill(pk_header, ffi.sizeof(pk_header))
+   local ethertype = ffi.cast('uint16_t*', p.data+12)[0]
+   if ethertype == 0xDD86 or ethertype == 0x0080 then
+      local startoffset = C.prepare_packet(p.data+14, p.length-14)
```

Comment:
Principle of least surprise: How come this app is inspecting the packet payload and requesting that the checksum be recalculated?

I have a few objections to this:

1. How about if the sender has already calculated the checksum? Then it is wasteful to ask for the device to repeat this work. (This will have overhead on the hypervisor in situations where the checksum is not offloaded onto the NIC.)
1. How about if the sender is supposed to pass packet payload verbatim? e.g. a Layer-2 device that should pass packets unmodified regardless of the validity of their Layer-3/Layer-4 checksums.
1. How does this really behave and are there some invariants that need to be taken care of? (For example does the packet need to already have a valid "header pseudo-checksum" stored in its checksum field?)

One way to take care of all these objections would be to skip the checksum offload feature and simply transmit packets verbatim. Then the other apps that are feeding the packets would need to take care of the checksums e.g. preserving the one that is already there, or calculating with our SIMD routines, or calculating a delta when updating headers. That may be unacceptable for other reasons but that would be an interesting discussion in its own right.